draft-ietf-taps-interface.md

```diff
@@ -1780,38 +1772,170 @@ that the other endpoint is done sending on a connection.
+also be used with datagram- or message-based protocols. In these cases,
+they add an additional transformation to further encode or encapsulate,
+and can potentially support packing multiple application-layer Messages
+into individual transport datagrams.
+
+## Defining Message Framers
+
+Applications can define a Message Framer by creating a object to represent a unique
+definition of a framing protocol. Event handlers for inbound and outbound data
+are registered on this object.
 
 ~~~
-Preconnection.DeframeWith(Deframer)
+framer := NewMessageFramer()
+framer.setEventHandlers()
+~~~
+
+## Adding Message Framers to Connections
 
-{messageData} := Deframer.Deframe(OctetStream)
+The Message Framer object can then be added to one or more Preconnections
+to run on top of transport protocols. Multiple Framers may be added.
+
+~~~
+Preconnection.AddFramer(framer)
+~~~
+
+When sending Messages, applications can add specific Message
+values to a MessageContext ({{message-props}}) that is intended for a Framer.
+This can be used, for example, to set the type of a Message for a TLV protocol.
+The namespace of values is custom for each unique Message Framer.
+
+~~~
+messageContext := NewMessageContext()
+messageContext.add(framer, key, value)
+Connection.Send(messageData, messageContext)
+~~~
+
+When an application receives a MessageContext in a Receive event,
+it can also look to see if a value was set by a specific Message Framer.
+
+~~~
+messageContext.get(framer, key) -> value
+~~~
+
+## Message Framer Lifetime
+
+When a Connection establishment attempt begins, an event is delivered to
+notify the framer that a new Connection is being created and has a new
+instance of a Message Framer. Similarly, a stop event is delivered when a
+Connection is being torn down.
+
+~~~
+FramerInstance -> Start()
+FramerInstance -> Stop()
 ~~~
 
+When Message Framer gets a start message, it sets up its state and then indicates
+to the connection that it is ready to handle sending Messages by calling `ready` on
+the FramerInstance. This call can be delayed in case the protocol needs to do some
+simple handshake exchange before handling application Messages.
+
+~~~
+instance.ready()
+~~~
+
+At any time if the Message Framer encounters a fatal error, it can also cause the Connection
+to fail and provide an error.
+
+~~~
+instance.failed(Error)
+~~~
+
+## Sender-side Message Framing {#send-framing}
+
+Message Framers define an event that gets called on an instance of a
+Framer to handle a new Message being sent by the application.
+
+~~~
+FramerInstance -> NewSentMessage<messageData, messageContext, endOfMessage>
+~~~
+
+Upon receiving this event, a Message Framer implementation is responsible for
+performing any necessary transformations and sending the resulting data to the next
+protocol. Implementations SHOULD ensure that there is a way to pass the original data
+through without copying to improve performance.
```

Comment:
While this is a valid comment, is this really something that belongs in the API?

Reply:
This is based on the discussion at IETF 104 that indicated we wanted to ensure that the API allows for 0-copy. That is an API semantic decision.